### PR TITLE
Elements API updates backport

### PR DESCRIPTION
**Pull request by MaggieCabrera**
This PR backports the following changes:

- https://github.com/WordPress/gutenberg/pull/40889
- https://github.com/WordPress/gutenberg/pull/41160
- https://github.com/WordPress/gutenberg/pull/40260
- https://github.com/WordPress/gutenberg/pull/41240
- https://github.com/WordPress/gutenberg/pull/41753
- https://github.com/WordPress/gutenberg/pull/41822
- https://github.com/WordPress/gutenberg/pull/41786
- https://github.com/WordPress/gutenberg/pull/41696
- https://github.com/WordPress/gutenberg/pull/41140
- https://github.com/WordPress/gutenberg/pull/42072
- https://github.com/WordPress/gutenberg/pull/42096
- https://github.com/WordPress/gutenberg/pull/41981
- https://github.com/WordPress/gutenberg/pull/42669
- https://github.com/WordPress/gutenberg/pull/42776 ⚠️ I don't think there's anything to backport on this one
- https://github.com/WordPress/gutenberg/pull/43088
- https://github.com/WordPress/gutenberg/pull/43167
- https://github.com/WordPress/gutenberg/pull/41335
- https://github.com/WordPress/gutenberg/pull/41446
- https://github.com/WordPress/gutenberg/pull/43988

Trac ticket: https://core.trac.wordpress.org/ticket/56467

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/class-wp-theme-json.php
Comment (ockham) [inline src/wp-includes/class-wp-theme-json.php:1334]: Does this need a `@since 6.1.0` PHPDoc?

File: src/wp-includes/class-wp-theme-json.php
Comment (ockham) [inline src/wp-includes/class-wp-theme-json.php:1394]: Does this need a `@since 6.1.0` PHPDoc?

File: src/wp-includes/class-wp-theme-json.php
Comment (cbravobernal) [inline src/wp-includes/class-wp-theme-json.php:364]: Should we add the changes in the comments?
`// @since 6.1.0 Added heading, button and caption elements`

**Comment by MaggieCabrera** [comment]
While trying to fix the tests I found a bug that wasn't caught in GB because the test is only on core, not in the GB repo. I'm going to try and patch that so we can add it to this one or we could "hack" the snapshot and then patch afterwards. I'm more inclined to do the former.

**Comment by MaggieCabrera** [comment]
[This is the fix](https://github.com/WordPress/gutenberg/pull/43988) for the issue that the tests uncovered.

Edit: merged and backported here too

**Comment by MaggieCabrera** [comment]
this only needs a review and check what's breaking the tests (they are all passing locally)

**Comment by SergeyBiryukov** [comment]
The tests previously [failed on PHP 5.6](https://github.com/WordPress/wordpress-develop/runs/8266559164?check_suite_focus=true#step:17:16) with a fatal error:
`Fatal error: Cannot use isset() on the result of an expression (you can use "null !== expression" instead) in src/wp-includes/class-wp-theme-json.php on line 1526`

0015d99 should resolve the issue. We already used `array_key_exists( $element, static::VALID_ELEMENT_PSEUDO_SELECTORS )` in one place, and that resolves the error in other instances in my testing.

**Review by SergeyBiryukov — APPROVED** [review]
This looks good to me 👍 Made some minor adjustments to documentation and unit tests for consistency with the rest of core. Going to merge shortly.

**Comment by SergeyBiryukov** [comment]
Thanks for the PR! Merged in [r54118](https://core.trac.wordpress.org/changeset/54118).

**Comment by ockham** [comment]
Thanks a lot, folks!

I was alerted that unfortunately, [r54118](https://core.trac.wordpress.org/changeset/54118) caused a fatal error on websites running Core `trunk` and a current version of the Gutenberg plugin 😕 

```
PHP Fatal error:  Cannot redeclare wp_add_global_styles_for_blocks() (previously declared in wp-includes/global-styles-and-settings.php:202) in wp-content/plugins/gutenberg/lib/compat/wordpress-6.1/get-global-styles-and-settings.php on line 12
```

We'll need to add a `! function_exists()` wrapper around the [function definition in Gutenberg](https://github.com/WordPress/gutenberg/blob/7f0a211da6603b336b75b75d8235a50bf547e6f3/lib/compat/wordpress-6.1/get-global-styles-and-settings.php#L8-L49) to avoid this.

Unfortunately, this won't fix the problem immediately: WP sites running Core `trunk` with the GB plugin active will continue to be affected until a new version of the plugin is released (as planned for next Thursday); or until the change is merged into GB `trunk` (if the site is running GB `trunk`). And finally, sites running Core `trunk` and an older version of GB (that has the unguarded `wp_add_global_styles_for_blocks` function definition) will continue to be affected 😕 

I'm not sure how to best cover the latter case. For mitigation, we might want to consider reverting [r54118](https://core.trac.wordpress.org/changeset/54118) (as much as it pains me to suggest that) -- at least until a GB version with the fix is released.

---

Going forward, we should probably put some strategies in place to avoid collisions like this:
- Require `wp_` function definitions in GB to be guarded with a `! function_exists()` check (and ideally enforce with a WPCS lint).
- Add a check in `wordpress-develop`'s CI to install and enable GB alongside Core (and run some basic tests to make sure it doesn’t fatal).

Anyway, this is something we'll discuss separately. I'll file one or two issues tomorrow.

**Comment by ockham** [comment]
Update: Unbeknownst to me, the `wp_` prefix has been changed to `gutenberg_` in the GB plugin: https://github.com/WordPress/gutenberg/pull/44052

**Comment by ockham** [comment]
Update #2: Let's hold off on the revert. We might release a Gutenberg point release with [this fix](https://github.com/WordPress/gutenberg/pull/44052) as a stop-gap.

**Pull request closed without merging**
